evm/vm/forks/sharding/log_handler.py

```diff
@@ -0,0 +1,117 @@
+import logging
+
+from eth_utils import (
+    to_dict,
+)
+
+
+def get_recent_block_hashes(w3, history_size):
+    block = w3.eth.getBlock('latest')
+    recent_hashes = [block['hash']]
+    # initialize the list of recent hashes
+    for _ in range(history_size - 1):
+        # break the loop if we hit the genesis block.
+        if block['number'] == 0:
+            break
+        block = w3.eth.getBlock(block['parentHash'])
+        recent_hashes.append(block['hash'])
+    reversed_recent_hashes = tuple(reversed(recent_hashes))
+    return reversed_recent_hashes
+
+
+def check_chain_head(w3, recent_block_hashes, history_size):
+    block = w3.eth.getBlock('latest')
+
+    new_block_hashes = []
+
+    for _ in range(history_size):
+        if block['hash'] in recent_block_hashes:
+            break
+        new_block_hashes.append(block['hash'])
+        block = w3.eth.getBlock(block['parentHash'])
+    else:
+        raise Exception('No common ancestor found for block: {0}'.format(block['hash']))
+
+    first_common_ancestor_idx = recent_block_hashes.index(block['hash'])
+
+    revoked_hashes = recent_block_hashes[first_common_ancestor_idx + 1:]
+
+    # reverse it to comply with the order of `self.recent_block_hashes`
+    reversed_new_block_hashes = tuple(reversed(new_block_hashes))
+
+    return revoked_hashes, reversed_new_block_hashes
+
+
+def preprocess_block_param(w3, block_param):
+    if isinstance(block_param, int):
+        return block_param
+    elif not isinstance(block_param, str):
+        raise ValueError("block parameter {0} must be int or str type".format(block_param))
+    current_block_number = w3.eth.blockNumber
+    if block_param == 'earliest':
+        return 0
+    elif block_param == 'latest':
+        return current_block_number
+    elif block_param == 'pending':
+        return current_block_number + 1
+    else:
+        raise ValueError(
+            "block parameter {0} in string must be one of earliest/latest/pending".format(
+                block_param,
+            )
+        )
+
+
+class LogHandler:
+
+    logger = logging.getLogger("evm.chain.sharding.LogHandler")
+
+    def __init__(self, w3, history_size=256):
+        self.history_size = history_size
```

Comment:
For `LogHandler` or higher levels (`ShardTracker`, `VMC`) APIs, is the `LogHandler.history_size` a *fixed number* of previous block hashes/headers capacity from the head block and *it is only set in initialization*?
If it's true, we may need a more flexible method to enable us to query for *older* block hashes (more than previous 256 block headers) *dynamically* during `guess_head` algorithm. The reason is that we want to verify as many collations as we can.